rclrs/src/lib.rs

```diff
@@ -119,6 +124,81 @@ pub fn create_node_builder(context: &Context, node_name: &str) -> NodeBuilder {
     Node::builder(context, node_name)
 }
 
+/// Extract non-ROS arguments from program's input arguments.
+///
+/// `args` is expected to be input arguments of the program (passed via [`env::args()`])
+/// According to rcl documentation ROS arguments are between `--ros-args` and `--` arguments.
+/// Everything beside is considered as non-ROS and left unparsed. Extracted non-ROS args are
+/// returned in order as they appear (see example below).
+///
+/// # Example
+/// ```
+/// let non_ros_args = rclrs::extract_non_ros_args(vec!["arg1", "--ros-args", "some", "args", "--", "arg2"]);
+/// assert_eq!(non_ros_args.len(), 2);
+/// assert_eq!(non_ros_args[0], "arg1");
+/// assert_eq!(non_ros_args[1], "arg2");
+/// ```
+pub fn extract_non_ros_args(
+    args: impl IntoIterator<Item = String>,
+) -> Result<Vec<String>, RclrsError> {
+    // SAFETY: No preconditions for this function
+    let mut rcl_arguments = unsafe { rcl_get_zero_initialized_arguments() };
+    // SAFETY: No preconditions for this function
+    let allocator = unsafe { rcutils_get_default_allocator() };
+
+    let cstring_args: Vec<CString> = args
+        .into_iter()
+        .map(|arg| {
+            CString::new(arg.as_str()).map_err(|err| RclrsError::StringContainsNul {
+                err,
+                s: arg.clone(),
+            })
+        })
+        .collect::<Result<_, _>>()?;
+    // Vector of pointers into cstring_args
+    let c_args: Vec<*const c_char> = cstring_args.iter().map(|arg| arg.as_ptr()).collect();
```

Comment:
Here is a code repetition from [context.rs](https://github.com/ros2-rust/ros2_rust/blob/main/rclrs/src/context.rs#L65). Initially I've isolated it to a function, but that's lead to double free error during UTs for some reason. Are you ok with this repetition? If not, could you please give me a hint how can I isolate this fragment?

Reply:
Yes, that's okay.

Reply:
To expand on that, if you try to make a function like
```rust
fn to_c_char_ptrs(args: &[String]) -> Result<Vec<*const c_char>, RclrsError> {
    let cstring_args: Vec<CString> = args
        .into_iter()
        .map(|arg| {
            CString::new(arg.as_str()).map_err(|err| RclrsError::StringContainsNul {
                err,
                s: arg.clone(),
            })
        })
        .collect::<Result<_, _>>()?;
    // Vector of pointers into cstring_args
    Ok(cstring_args.iter().map(|arg| arg.as_ptr()).collect())
}
```
then the returned pointers will be dangling (the `cstring_args` they're pointing into will be gone after the function returns). 

We could either do something clever where we return the `cstring_args` along with the `c_args` in one struct, or we could make it a macro instead of a function. But leaving it like it is is fine for now, we can do that in a separate PR.